lib/models/package-info-cache.js

```diff
@@ -0,0 +1,1049 @@
+      this.ui.writeWarnLine('Top level errors:');
+      rootPath = this.realPath || '';
+    } else {
+      this.ui.writeWarnLine(`${obj.realPath}:`);
+      rootPath = obj.realPath;
+    }
+
+    let errorList = obj.errors;
+
+    errorList.getErrors().forEach(errorEntry => {
+      switch (errorEntry.type) {
+        case ERROR_PACKAGE_DIR_MISSING:
+          this.ui.writeWarnLine(`   Missing package directory`);
+          break;
+        case ERROR_PACKAGE_JSON_MISSING:
+          this.ui.writeWarnLine(`   Missing package.json file`);
+          break;
+        case ERROR_PACKAGE_JSON_PARSE:
+          this.ui.writeWarnLine(`   Error parsing package.json file`);
+          break;
+        case ERROR_EMBER_ADDON_MAIN_MISSING:
+          this.ui.writeWarnLine(
+            `   Missing ember-addon 'main' file at relative path '${path.relative(
+              rootPath,
+              errorEntry.data
+            )}'`
+          );
+          break;
+        case ERROR_DEPENDENCIES_MISSING:
+          if (errorEntry.data.length === 1) {
+            this.ui.writeWarnLine(
+              `   Missing dependency '${errorEntry.data[0]}'`
+            );
+          } else {
+            this.ui.writeWarnLine(`   Missing dependencies:`);
+            errorEntry.data.forEach(dependencyName => {
+              this.ui.writeWarnLine(`      ${dependencyName}`);
+            });
+          }
+          break;
+        case ERROR_NODEMODULES_ENTRY_MISSING:
+          this.ui.writeWarnLine(`   Missing node_modules entry '[${errorEntry.data}'`);
+          break;
+      }
+    });
+  }
+
+  /**
+   * Process the root directory of a project, given a
+   * Project object (we need the object in order to find the internal addons).
+   * _readPackage takes care of the general processing of the root directory
+   * and common locations for addons, filling the cache with each. Once it
+   * returns, we take care of the locations for addons that are specific to
+   * projects, not other packages (e.g. internal addons, cli root).
+   *
+   * Once all the project processing is done, go back through all cache entries
+   * to create references between the packageInfo objects.
+   *
+   * @public
+   * @method loadProject
+   * @return {PackageInfo} the PackageInfo object for the given Project object.
+   * Note that if the project path is already in the cache, that will be returned.
+   * No copy is made.
+   */
+  loadProject(projectObj) {
+    let val = this._readPackage(projectObj.root, projectObj.pkg);
+
+    // NOTE: the returned val may contain errors, or may contain
+    // other packages that have errors. We will try to process
+    // things anyway.
+    if (!val.processed) {
+      this.projects.push(projectObj);
+
+      // projects are a bit different than standard addons, in that they have
+      // possibly a CLI addon and internal addons. Add those now.
+      let pkgInfo = val;
+
+      pkgInfo.project = projectObj;
```

Comment:
why are we assigning `project` property only in `loadProject`?

Reply:
Because the PackageInfo itself is general - it is used for any package.json. The only place where we actually know we have a project is in loadProject(). That's also why project-specific stuff is done in loadProject (like cli.root and supportedInternalAddonPaths) rather than _readPackage. Make sense?